JobScheduler.js

```diff
@@ -1,84 +1,215 @@
-      removeJobWorker(worker.id);
-      addJobWorker();
-    }, createWorkerDelay);
-  });
-  process.on('SIGTERM', notifyShutDown);
-  process.on('SIGINT', notifyShutDown);
-  //https://github.com/nodejs/node-v0.x-archive/issues/5054
-} else {
-  workerType = `Worker - ${cluster.worker.id} - ${process.pid} `;
-  logger.info(`Starting Service Fabrik Batch Job worker: ${cluster.worker.id} - ${process.pid}  @${new Date()}`);
-  require('./lib/jobs');
-  require('./lib/fabrik');
-}
+      this.removeJobWorker(worker.id);
+      this.addJobWorker();
+    }, CONST.JOB_SCHEDULER.WORKER_CREATE_DELAY);
+  }
 
-function addJobWorker() {
-  workerCount++;
-  jobWorkers.push(cluster.fork({
-    job: 1,
-    worker: workerCount
-  }).id);
-}
+  placeSchedulerInMaintenance() {
+    this.serviceFabrikInMaintenance = true;
+    _.each(this.jobWorkers, (id, key) => {
+      logger.info(`+-> message From -> ${this.workerType} - To worker - ${id} - ${key}-${JSON.stringify(cluster.workers[key])}}`);
+      cluster.workers[key].send(CONST.TOPIC.APP_SHUTTING_DOWN);
+    });
+    this.workerCount = 0;
+    this.jobWorkers = [];
+    this.ensureSystemNotInMainenanceThenInitMaster();
+  }
 
-function removeJobWorker(id) {
-  jobWorkers.splice(jobWorkers.indexOf(id), 1);
-}
+  addJobWorker() {
+    logger.info('adding worker : ', !this.serviceFabrikInMaintenance);
+    if (!this.serviceFabrikInMaintenance) {
+      this.workerCount++;
+      const worker = cluster.fork({
+        job: 1,
+        worker: this.workerCount
+      });
+      this.jobWorkers.push(worker.id);
+    }
+  }
 
-function notifyShutDown() {
-  logger.info(`ServiceFabrik Batch Job ${workerType} shutting down shortly...`);
-  pubsub.publish(CONST.TOPIC.APP_SHUTTING_DOWN);
-  const waitBeforeShutdown = process.env.NODE_ENV === 'test' ? 0 : 5000;
-  setTimeout(() => {
-    logger.info(`ServiceFabrik Batch Job ${workerType} shutdown complete`);
-    return process.env.NODE_ENV !== 'test' ? process.exit(2) : '';
-  }, waitBeforeShutdown);
-}
+  removeJobWorker(id) {
+    this.workerCount--;
+    this.jobWorkers.splice(this.jobWorkers.indexOf(id), 1);
+  }
+
+  shutDown() {
+    logger.info(`ServiceFabrik Batch Job ${this.workerType} shutting down shortly...`);
+    pubsub.publish(CONST.TOPIC.APP_SHUTTING_DOWN);
+    const waitBeforeShutdown = CONST.JOB_SCHEDULER.SHUTDOWN_WAIT_TIME;
+    setTimeout(() => {
+      logger.info(`ServiceFabrik Batch Job ${this.workerType} shutdown complete`);
+      process.exit(2);
+    }, waitBeforeShutdown);
+  }
 
-process.on('unhandledRejection', (reason, p) => {
-  if (reason && reason instanceof errors.DBUnavailable) {
-    logger.error('DB unavailable. shutting down app');
-    notifyShutDown();
-  } else {
-    logger.error('Unhandled Rejection at:', p, 'reason:', reason);
+  pollMaintenanceStatus() {
+    const checkMaintenanceStatus = (resolve, reject) => {
+      return maintenanceManager
```

Comment:
We agreed that in fresh landscape maintenance_start / end APIs will not be getting invoked. Once scheduler starts it errors out because it cant connect to DB & it kills itself since it [could not determine maintenance status](https://github.com/jagadish-kb/service-fabrik-broker/blob/master/JobScheduler.js#L191) & monit keeps brining back scheduler process. Once DB is provisioned the next moment, DB ConnectionManager connects and I guess it gets back maintenance status as null and then workers get started. Haven't tested this scenario though, but this is how it should work :-)

Reply:
Ok , got it.